Add enrich-kafka

oguzhanunlu opened the pull request:
Let's get the review on the initial implementation started!

This implementation is based on [fs2-kafka](https://github.com/fd4s/fs2-kafka) library. I followed the conventions of enrich-kinesis & enrich-pubsub. All of the [consumer configs](https://kafka.apache.org/documentation/#consumerconfigs) and [producer configs ](https://kafka.apache.org/documentation/#producerconfigs) can be set at runtime.

Together with stream-collector-kafka, I had a successful smoke test on localhost.

I'll focus on optimizing the throughput after passing the initial review phase. I expect our reference config to have different default values for some consumer and producer configs only after working on throughput optimization.

Next steps
-----
- [x] reach a consensus on whether auto-commit should be allowed or not
- [x] benchmark against stream-enrich-kafka
- [x] prepare an integration test ( #686 )
- [x] write unit test for attributesFromFields
- [x] Initial Azure test to re-confirm compatibility

benjben commented on modules/kafka/src/main/resources/application.conf:
```diff
@@ -0,0 +1,71 @@
+    "topicName": "collector-payloads"
+    "bootstrapServers": "localhost:9092"
+    "consumerConf": {
+      "enable.auto.commit": "false"
```
```suggestion
      "enable.auto.commit": "false"
      "group.id": "enrich"
```
Group id is [required](https://docs.confluent.io/platform/current/installation/configuration/consumer-configs.html#consumerconfigs_group.id) given that we use `.subscribe()`.
Also, it would be good to make it appear in the extended example so that users can see what the default value is.

benjben commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Source.scala:
```diff
@@ -0,0 +1,45 @@
+      ConsumerSettings[F, String, Array[Byte]]
+        .withBootstrapServers(input.bootstrapServers)
+        .withEnableAutoCommit(false)
+        .withProperties(input.consumerConf)
```
At first I wanted to suggest that :
```suggestion
        .withProperties(input.consumerConf)
        .withBlocker(blocker)
```
but the documentation says:
```
Because the underlying Java consumer is not thread-safe, the blocker must be single-threaded. If in doubt, leave this unset so that a default single-threaded blocker will be provided.
```
and we don't want to make our blocking thread pool single-threaded so I think that we should leave it as is.

oguzhanunlu replied:
yeah, I also went through the same thinking and ended up leaving it as is

benjben reviewed (changes requested):
Looks great overall @oguzhanunlu !

oguzhanunlu commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Source.scala:
```diff
@@ -0,0 +1,45 @@
+    val consumerSettings =
+      ConsumerSettings[F, String, Array[Byte]]
+        .withBootstrapServers(input.bootstrapServers)
+        .withEnableAutoCommit(false)
```
@benjben it is possible to override disabled auto-commit via providing the related consumer config, do we have a consensus on whether it should be allowed or not? I can swap L37 with L38 to prevent enabling auto-commits.

benjben replied:
I'm in favor of always disabling it, to have at least once processing and make it impossible for enrich to lose data

oguzhanunlu commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Sink.scala:
```diff
@@ -0,0 +1,65 @@
+      case o => Resource.eval(Concurrent[F].raiseError(new IllegalArgumentException(s"Output $o is not Kafka")))
+    }
+
+  private def mkProducer[F[_]: ConcurrentEffect: ContextShift](output: Output.Kafka): Resource[F, KafkaProducer[F, String, Array[Byte]]] = {
```
For the sake of simplification, I'm in favor of enforcing. I'll put related producer config properties in the extended example.

oguzhanunlu commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Main.scala:
```diff
@@ -0,0 +1,71 @@
+ */
+
+package com.snowplowanalytics.snowplow.enrich.kafka
+
+import cats._
+import cats.implicits._
+import cats.effect.{ExitCode, IO, IOApp, Resource, SyncIO}
+import com.snowplowanalytics.snowplow.enrich.common.fs2.{Run, Telemetry}
+import fs2.kafka.CommittableConsumerRecord
+
+import scala.concurrent.ExecutionContext
+
+import java.util.concurrent.{Executors, TimeUnit}
+import com.snowplowanalytics.snowplow.enrich.kafka.generated.BuildInfo
```
Yeah imports aren't well organized but scalafmt is happy, modifying scalafmt would lead to a giant PR here so I'll manually fix it for now. I'd prefer to fix imports in another PR

benjben replied:
You don't need to modify scalafmt to reorganize the imports, scalafmt doesn't care much about them (with our configuration)

oguzhanunlu replied:
fixed them manually for today

oguzhanunlu commented on modules/kafka/src/main/resources/application.conf:
```diff
@@ -0,0 +1,71 @@
+    "pii": {
+      "type": "Kafka"
+      "topicName": ""
+      "bootstrapServers": "localhost:9092"
```
I put a 👍 but not 100% sure. Could you explain this further @benjben ?

benjben replied:
Oops sorry there was a mistake in my sentence (I had updated my initial one), it should have been
```
This can't be a default value for that
```
The reason is that default values have to work, and here if we specify just a `topicName` but no `bootstrapServers` it won't work (unless with local brokers but that's only for testing)

oguzhanunlu replied:
I made `pii.bootstrapServers` equal to `""` but this makes the app fail with `org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers` when we provide topicName without bootstrapServers. -- I'll improve config decoder to beautify this

If this is the behaviour we want to have, I'll empty `good.bootstrapServers` and `bad.bootstrapServers` as well in application.conf cc @benjben 

Please let me know if I misunderstand your point

benjben replied:
Indeed `bootstrapServers` should be removed from `application.conf` for good and bad, as well as `topicName`, as only values that will always work should be in `application.conf`.

For `pii`:
- We want to specify some default values for `type` and `producerConf`
- To specify default values we need `pii` block to exist in `application.conf`
- If `pii` block exists, it will be parsed and needs to be valid in terms of [this class](https://github.com/snowplow/enrich/blob/enrich-kafka/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/io.scala#L214-L219)
- To be valid, all the non-optional class members have to be defined
- So `bootstrapServers` and `topicName` have to be there
- What we can do is to set them to empty string, and then check and if they are both empty we should remove pii output, exactly like [this is done for Kinesis and PubSub](https://github.com/snowplow/enrich/blob/master/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ConfigFile.scala#L75-L80)
- Then [here](https://github.com/snowplow/enrich/blob/enrich-kafka/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/io.scala#L263-L264) you need to make the parsing fail if only one of `bootstrapServers` and `topicName` is not empty

oguzhanunlu replied:
thanks @benjben ! I just pushed a separate commit to address this, I previously implemented pii removal [here](https://github.com/snowplow/enrich/blob/enrich-kafka/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ConfigFile.scala#L81-L83) and updated output decoder [here](https://github.com/snowplow/enrich/blob/enrich-kafka/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/io.scala#L263-L264) to fail if one of them is empty

oguzhanunlu replied:
I needed to update output decoder to fail only if one of both fields is empty, done in a separate commit

oguzhanunlu commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Sink.scala:
```diff
@@ -0,0 +1,65 @@
+    val producerSettings =
+      ProducerSettings[F, String, Array[Byte]]
+        .withBootstrapServers(output.bootstrapServers)
+        .withProperties(output.producerConf)
```
Resolving as we are on the same page about not providing our own blocker

benjben replied:
For the source ! For the sink we should provide it as there is no constraint that the blocking thread pool should be single-threaded

oguzhanunlu replied:
right! thanks for the correction! I'll provide it!

oguzhanunlu replied:
done!

oguzhanunlu commented on project/BuildSettings.scala:
```diff
@@ -99,6 +99,14 @@ object BuildSettings {
+  lazy val kafkaProjectSettings = projectSettings ++ Seq(
+    name := "snowplow-enrich-kafka",
+    moduleName := "snowplow-enrich-kafka",
+    description := "High-performance streaming enrich app with Kafka source, built on top of functional streams",
```
thanks Ben, I added a separate commit to reword project descriptions of new enrich apps, since we can have file source for them as well

oguzhanunlu commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ParsedConfigs.scala:
```diff
@@ -129,6 +131,12 @@ object ParsedConfigs {
+          case (s, _) =>
+            s == partitionKey
+        }
+        attributesFromFields(fields)
```
I have a separate commit applying DRY here without changing a signature, how does it look?

benjben replied:
Yes that looks good! It would be great to add a unit test to `attributesFromFields` function

oguzhanunlu replied:
will do, thanks Ben!

oguzhanunlu commented:
hey @benjben , I addressed your feedback in separate commits but I'm not sure about the 2 unresolved points above, could you have another look please? Thanks in advance.

a note on replacing partitionKey with headers: our AttributedData class isn't capable of carrying partition key & attributes separately hence the decision to replace partition key with record headers for Kafka. Neither Kinesis nor PubSub requires both partition key & attributes. I can have a look at extending AttributedData to support both concepts if you think it is worth it (atm record keys are random UUIDs, as if no partition key is given)

benjben commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Sink.scala:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  private def toProducerRecord(topicName: String, record: AttributedData[Array[Byte]]): ProducerRecord[String, Array[Byte]] =
+    ProducerRecord(topicName, UUID.randomUUID().toString, record.data)
```
We don't need to generate a UUID as Kafka will automatically distribute the records randomly across partitions for us if  no partition and no key is set:
```suggestion
    ProducerRecord(topicName, null, record.data)
```
But we want to be able to specify a partition key in the config. If we specify a key in the `ProducerRecord`, it is automatically used as a partition key:
```
If no partition is provided, and the ProducerRecord has a key, the producer takes the hash of the key modulo the number of partitions. The resulting number from that calculation is the partition that the producer will use.
```
So we want to set the key of `ProducerRecord` to the value of the field set as the `partitionKey` in the config, if defined.

So to answer [your question](https://github.com/snowplow/enrich/pull/659#issuecomment-1201435017) indeed we need to extend `AttributedData` to have both a partition key value and some attributes.

oguzhanunlu replied:
thanks Ben, I'll extend AttributedData to carry both

oguzhanunlu replied:
addressed in a separate commit

oguzhanunlu commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/io.scala:
```diff
@@ -202,6 +210,14 @@ object io {
+    case class Kafka private (
+      topicName: String,
+      bootstrapServers: String,
+      headers: Option[Set[String]],
```
definitely! thanks 👍

oguzhanunlu commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Main.scala:
```diff
@@ -0,0 +1,81 @@
+      None
+    )
+
+  private def checkpoint[F[_]: Applicative](records: List[CommittableConsumerRecord[F, String, Array[Byte]]]): F[Unit] =
+    if (records.isEmpty) Applicative[F].unit
+    else
+      records
+        .reduceLeft { (biggest, current) =>
+          if (current.record.offset > biggest.record.offset) current
+          else biggest
+        }
+        .offset
+        .commit
+}
```
very good point, thanks @benjben !

oguzhanunlu replied:
done in a separate commit

oguzhanunlu commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/io.scala:
```diff
@@ -244,6 +260,8 @@ object io {
     implicit val outputDecoder: Decoder[Output] =
       deriveConfiguredDecoder[Output]
         .emap {
+          case k: Kafka if k.topicName.isEmpty && k.bootstrapServers.nonEmpty => "topicName can't be empty".asLeft
+          case k: Kafka if k.topicName.nonEmpty && k.bootstrapServers.isEmpty => "bootstrapServers can't be empty".asLeft
```
nice shortcut for the same logic, thanks!

oguzhanunlu commented:
hey @benjben , I addressed your latest feedback in the latest 3 commits, could you please have another look please?

istreeter commented on modules/common-fs2/src/test/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/test/TestEnvironment.scala:
```diff
@@ -167,6 +167,8 @@ object TestEnvironment extends CatsIO {
                       metrics,
                       metadata,
                       None,
+                      _ => "",
+                      _ => "",
```
I guess this could be any string.  Why did you choose the empty string instead of some placeholder?  If you used a distinctive string (e.g. `test_good_partition_key`) then you could assert that Enrich sets the correct key.

oguzhanunlu replied:
no reason really, that's a great idea, thanks!

istreeter commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Environment.scala:
```diff
@@ -114,6 +114,8 @@ final case class Environment[F[_], A](
   metrics: Metrics[F],
   metadata: Metadata[F],
   assetsUpdatePeriod: Option[FiniteDuration],
+  goodPartitionKey: EnrichedEvent => String,
+  piiPartitionKey: EnrichedEvent => String,
```
Update the scaladoc above the class declaration.

istreeter commented on modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Main.scala:
```diff
@@ -0,0 +1,81 @@
+      records
+        .groupBy(_.record.partition)
+        .mapValues { records =>
+          records.reduceLeft { (biggest, current) =>
+            if (current.record.offset > biggest.record.offset) current
+            else biggest
+          }
```
Does this also work: `records.maxBy(_.record.offset)` ?

oguzhanunlu replied:
it does! thanks for the shortcut

istreeter commented on config/config.kafka.extended.hocon:
```diff
@@ -0,0 +1,249 @@
+
+      # Optional, Kafka producer configuration
+      # See https://kafka.apache.org/documentation/#producerconfigs for all properties
+      "producerConf": {
```
I can imagine wanting to override a setting in my config.  As a user, I would hope this is a legitimate configuration file:

```
  "output": {
    "good": {
      // other fields go here
      "producerConf": {
        "buffer.memory": 1000000
      }
    }
  }
```

But what happens with that config?  Does my map of a single key-value _override_ the `producerConf` map in the `application.conf` or does it _append_ to the map?  If it overrides the map then it's a real problem, because then my serializers are not set properly.

oguzhanunlu replied:
it doesn't override, user-provided config is merged with the config from `application.conf` and atm we have a single key-value pair by default (`acks: all`) in `application.conf`

so your config is appended and it is possible to override `acks` setting as well (similar to consumer not allowing to override auto commit setting, we can enforce `acks` as well)

oguzhanunlu replied:
I noticed fs2-kafka not setting key serializer/deserializer as expected, I'll send a new commit to enforce key-value serializer/deserializer which will also disable our users to change them

istreeter replied:
Now that you are adding the serializer/deserializer using the `withProperties` method in `Source.scala`... does this mean you can remove them from the `application.conf` file and from the example config file?

oguzhanunlu replied:
yes! I'll include remove them from example config file

istreeter reviewed (commented):
This is looking really good @oguzhanunlu

oguzhanunlu commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ParsedConfigs.scala:
```diff
@@ -111,31 +115,44 @@ object ParsedConfigs {
+      case _ => _ => ""
+    }
+
+  private def partitionKeyFromFields(partitionKey: String)(ee: EnrichedEvent): String =
+    ParsedConfigs.enrichedFieldsMap.get(partitionKey).fold("") { f =>
+      Option(f.get(ee)).fold("")(_.toString)
+    }
```
this is a great suggestion, thanks!

oguzhanunlu commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Telemetry.scala:
```diff
@@ -125,6 +125,7 @@ object Telemetry {
   object Cloud {
     case object Aws extends Cloud
     case object Gcp extends Cloud
+    case object Agnostic extends Cloud
```
I checked with @stanch , we'll match what stream-collector [does](https://github.com/snowplow/stream-collector/blob/master/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/telemetry/TelemetryAkkaService.scala#L123), i.e. use `None` instead of `Some(Agnostic)`

Thanks for the heads-up!

oguzhanunlu commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/io.scala:
```diff
@@ -244,6 +261,8 @@ object io {
     implicit val outputDecoder: Decoder[Output] =
       deriveConfiguredDecoder[Output]
         .emap {
+          case Kafka(topicName, bootstrapServers, _, _, _) if topicName.isEmpty ^ bootstrapServers.isEmpty =>
+            "Both topicName and bootstrapServers have to be set".asLeft
```
just pushed a separate commit to make input decoder like output decoder

I'll update both encoder and decoder if we decide to separate error messages

istreeter replied:
Good that you made them the same now!!  But to sorry to say, I preferred it how it was previously for the input decoder with separate error messages.

I disagree with [Ben's comment](https://github.com/snowplow/enrich/pull/659#discussion_r937721747) where he said:

> They can be both empty (for pii). What we don't want is that only one of them is set but not the other.

They cannot both be empty!  Even for pii.  For pii out, the entire `Output` field is allowed to be undefined, but if it is defined then both `topicName` and `bootstrapServers` must be non-empty.

istreeter replied:
Ignore this comment above ☝️, I was wrong.  I took a closer look at the code and now I understand why it's possible for topicName and bootstrapServers to be empty for pii

istreeter commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ParsedConfigs.scala:
```diff
@@ -111,31 +115,45 @@ object ParsedConfigs {
-      case (k, f) =>
-        Option(f.get(ee)).map(v => k -> v.toString)
+  private[config] def partitionKeyFromFields(partitionKey: String): EnrichedEvent => String =
+    ParsedConfigs.enrichedFieldsMap.get(partitionKey).fold[EnrichedEvent => String](_ => "") { f => ee =>
```
What would it mean if `ParsedConfigs.enrichedFieldsMap` is _missing_ the key `partitionKey`.  Would this mean the user has provided an invalid hocon file? Do you have any validation anywhere to prevent this happening?  I would think if the map is missing the partition key, then it would be better to raise an exception instead of returning `_ => ""`

oguzhanunlu replied:
> Would this mean the user has provided an invalid hocon file?

yes and I don't have any validation to prevent it atm. I'll improve output decoder to make sure enrich-kafka starts with a valid partitionKey and a valid set of headers so that we don't need to raise an exception

istreeter commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/AttributedData.scala:
```diff
@@ -12,5 +12,15 @@
+ */
+final case class AttributedData[A](
+  data: A,
+  partitionKey: String,
```
This is a nice change, to make the partition key a field of attributed data.

However, in the master branch we currently have [this code](https://github.com/snowplow/enrich/blob/3.2.3/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L172-L176) for kinesis, which takes the partition key from the attributes.  Do you think we should update that kinesis code to make use of this new partitionKey field?

oguzhanunlu replied:
I think we should update it. I intentionally left it out, not to touch enrich-kinesis in this PR. I was planning to open another PR after this one is merged. Would you prefer to address that in this PR?

oguzhanunlu replied:
addressed in this PR

istreeter reviewed (commented):
Hey @oguzhanunlu this is really looking very good.

This is not a thorough code review (sorry!); I just scribbled down a couple of thoughts as I browsed through the code.

istreeter commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ParsedConfigs.scala:
```diff
@@ -111,31 +115,45 @@ object ParsedConfigs {
-        Option(f.get(ee)).map(v => k -> v.toString)
+  private[config] def partitionKeyFromFields(partitionKey: String): EnrichedEvent => String =
+    ParsedConfigs.enrichedFieldsMap.get(partitionKey).fold[EnrichedEvent => String](_ => "") { f => ee =>
+      Option(f.get(ee)).fold("")(_.toString)
```
Interesting here that you return `""` if the event is missing the field.  Contrast with what happens in enrich-kinesis where [a random UUID is used as the partition key](https://github.com/snowplow/enrich/blob/3.2.3/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L175) if a fallback is needed.  I guess the point of the UUID is to prevent a "hot" shard that could receive a disproportionate fraction of the events.

oguzhanunlu replied:
From kafka [docs](https://kafka.apache.org/0102/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html): 
`If neither key nor partition is present a partition will be assigned in a round-robin fashion. `

so we don't need a random UUID as we don't specify key and partition

istreeter replied:
But you are setting a partition key.  You're setting it as `""`, which I believe Kafka treats as a valid partition key.  This is not the same thing as not setting a partition key.

I'm only like 80% sure about this, so please call me out if I'm wrong!

oguzhanunlu replied:
I think you're right! the same doc explains that

`If no partition is specified but a key is present a partition will be chosen using a hash of the key.`

so we'll be hitting the same partition by hashing empty string -- very important catch, thanks! I'll use a random UUID here
